include/tests/propagator.hpp

```diff
@@ -0,0 +1,167 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*! @file
+ * @brief A Propagator class to manage the loop for each the timestep decoupled of the tests
+ *
+ * @author Sebastian Keller <sebastian.f.keller@gmail.com>
+ * @author Jose A. Escartin <ja.escartin@gmail.com>
+ */
+
+#pragma once
+
+#include <iostream>
+
+#include "../Timer.hpp"
+#include "../Task.hpp"
+#include "../ParticlesData.hpp"
+
+#include "cstone/domain/domain.hpp"
+
+#include "sph/timestep.hpp"
+
+using namespace cstone;
+using namespace sphexa;
+using namespace sphexa::sph;
+
+template <class I, class T, class Accelerator = CpuTag>
+class Propagator
+{
+
+private:
+
+    Domain<I, T, Accelerator>& domain;                          //
+    ParticlesData<T, I>&       d;                               //
+
+    MasterProcessTimer         timer;                           //
+
+    const size_t               nTasks;                          //
+    const size_t               ngmax;                           //
+    const size_t               ng0;                             //
+
+    TaskList                   taskList;                        //
+
+    const bool                 gravity;                         //
+
+public:
+
+    Propagator(
+        Domain<I, T, Accelerator>& domain,                      //
+        ParticlesData<T, I>&       d,                           //
+        std::ostream&              output,                      //
+        const size_t               nTasks,                      //
+        const size_t               ngmax,                       //
+        const size_t               ng0,                         //
+        const bool                 gravity)                     //
+        : domain(domain)
+        , d(d)
+        , timer(output, d.rank)
+        , nTasks(nTasks)
+        , ngmax(ngmax)
+        , ng0(ng0)
+        , taskList(0, domain.nParticles(), nTasks, ngmax, ng0)
+        , gravity(gravity)
+    {}
+
+    float duration(){
+        return timer.duration();
+    }
+
+    void start(){
+        timer.start();
+    }
+
+    void stop(){
+        timer.stop();
+    }
+
+    void step(const std::string &name){
+         timer.step(name);
+    }
+
+    size_t neighbors(){
+        return neighborsSum(taskList.tasks);
+    }
```

Comment:
The back-and-force between timing inside the propagator and outside it adds a lot of clutter.
Do you think you could move all the timing inside the propagator and not time the file I/O, such that
the file I/O only is part of the final runtime, but not the iteration steps?